src/pybind/mgr/smart/module.py

```diff
@@ -1,33 +1,128 @@
 
 """
-Pulling smart data from OSD
+Pulling SMART data from OSD
 """
 
 import json
 from mgr_module import MgrModule, CommandResult
+import rados
+import time
+
+pool_name = "smart_data"
 
 
 class Module(MgrModule):
     COMMANDS = [
+        {
+            "cmd": "osd smart scrape "
+                   "name=osd_id,type=CephString,req=true",
+            "desc": "Scraping SMART data of osd<osd_id>",
+            "perm": "r"
+        },
+        {
+            "cmd": "osd smart scrape all",
+            "desc": "Scraping SMART data of all OSDs",
+            "perm": "r"
+        },
         {
             "cmd": "osd smart get "
                    "name=osd_id,type=CephString,req=true",
-            "desc": "Get smart data for osd.id",
+            "desc": "Get SMART data for osd_id",
+            "perm": "r"
+        },
+        {
+            "cmd": "osd smart get all",
+            "desc": "Get SMART data for all OSDs",
+            "perm": "r"
+        },
+        {
+            "cmd": "pool test",
+            "desc": "Just a pool test",
             "perm": "r"
         },
     ]
 
+    def open_connection(self):
+        # looking for the right pool
+        # TODO open pool, and if it fails - create and open
+        pools = self.rados.list_pools()
+        is_pool = False
+        for pool in pools:
+            self.log.debug('pool name is %s' % pool)
+            if pool == pool_name:
+                is_pool = True
+                break
+
+        # TODO make pool name configurable
+        if not is_pool:
+            self.rados.create_pool(pool_name)
+            self.log.debug('create %s pool' % pool_name)
+
+        ioctx = self.rados.open_ioctx(pool_name)
+
+        return (ioctx)
+
+    def handle_osd_smart_scrape_all(self):
+        osdmap = self.get("osd_map")
+        assert osdmap is not None
+
+        ioctx = self.open_connection()
+
+        for osd in osdmap['osds']:
+            osd_id = osd['osd']
+            self.do_osd_smart_scrape(str(osd_id), ioctx)
+
+        ioctx.close()
+        return (0, "", "")
+
+    def handle_osd_smart_scrape(self, osd_id):
+        ioctx = self.open_connection()
+        self.do_osd_smart_scrape(osd_id, ioctx)
+        ioctx.close()
+        return (0, "", "")
+
+    def do_osd_smart_scrape(self, osd_id, ioctx):
+        # running 'tell' command on OSD, retrieving smartctl results
+        result = CommandResult('')
+        self.send_command(result, 'osd', osd_id, json.dumps({
+            'prefix': 'smart',
+            'format': 'json',
+        }), '')
+        r, outb, outs = result.wait()
+        smart_data = json.loads(outb)
+
+        osd_host = self.get_metadata('osd', osd_id)['hostname']
```

Comment:
+        osd_host = self.get_metadata('osd', osd_id).get('hostname')
if not osd_host:
   return

Reply:
added; should we throw an exception?

Reply:
Better to just log someting (at debug level probably?) and continue.  Remember thi sis going to loop over 100s or 1000s of devices, so we can't just stop when we hit one that looks weird.